utils/zerovec/src/map/hashmap.rs

```diff
@@ -0,0 +1,280 @@
+                    if let Some(v) = occupied.get_mut(*displacement_idx) {
+                        *v = true;
+                    }
+                    if let Some(v) = reverse_mapping.get_mut(*displacement_idx) {
+                        *v = *original_idx;
+                    }
+                }
+                break;
+            }
+            start = end;
+        }
+
+        Self {
+            displacements: FlexZeroVec::Owned(FlexZeroVecOwned::from_iter(
+                displacements.into_iter(),
+            )),
+            reverse_mapping: FlexZeroVec::Owned(FlexZeroVecOwned::from_iter(
+                reverse_mapping.into_iter(),
+            )),
+        }
+    }
+
+    #[inline]
+    pub fn index<K>(&self, k: &K) -> Option<usize>
+    where
+        K: Hash,
+    {
+        let l1 = compute_hash(0, k, self.displacements.len());
+
+        #[allow(clippy::unwrap_used)] // l1 is in 0..self.displacements.len()
+        let seed = self.displacements.get(l1).unwrap();
+        if seed == 0 {
+            None
+        } else {
+            let hash = compute_hash(seed as u32, k, self.displacements.len());
+            self.reverse_mapping.get(hash).map(|i| i as usize)
+        }
+    }
+}
+
+pub struct ZeroHashMapStatic<'a, K, V>
+where
+    K: ZeroMapKV<'a> + ?Sized,
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    index: HashIndex<'a>,
+    keys: K::Container,
+    values: V::Container,
+}
+
+impl<'a, K, V> ZeroHashMapStatic<'a, K, V>
+where
+    K: ZeroMapKV<'a> + ?Sized,
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    pub fn len(&self) -> usize {
+        self.values.zvl_len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl<'a, K, V> ZeroHashMapStatic<'a, K, V>
+where
+    K: ZeroMapKV<'a> + ?Sized + Hash + 'static + Eq,
+    V: ZeroMapKV<'a> + ?Sized,
+{
+    #[inline]
+    pub fn get<'b>(&'b self, key: &'b K) -> Option<&'b V::GetType> {
+        self.index
+            .index(&key)
+            .and_then(|i| {
+                #[allow(clippy::unwrap_used)] // i is in 0..self.keys.len()
+                K::Container::zvl_get_as_t(self.keys.zvl_get(i).unwrap(), |k| key.eq(k)).then(|| i)
+            })
+            .and_then(|i| self.values.zvl_get(i))
+    }
+
+    /// Build a [`ZeroHashMapStatic`] from an iterator returning (K, V) tuples.
+    ///
+    /// # Example
+    /// ```
+    /// use zerovec::ZeroHashMapStatic;
+    ///
+    /// let kv: Vec<(i32, &str)> = vec![(1,"a"), (2, "b"),(3, "c"),(4 , "d")];
+    /// let hashmap: ZeroHashMapStatic<i32, str> = ZeroHashMapStatic::from_exact_iter(kv.into_iter());
+    /// assert_eq!(hashmap.get(&1), Some("a"));
+    /// assert_eq!(hashmap.get(&2), Some("b"));
+    /// assert_eq!(hashmap.get(&3), Some("c"));
+    /// assert_eq!(hashmap.get(&4), Some("d"));
+    /// ```
+    pub fn from_exact_iter<A, B, I>(iter: I) -> Self
+    where
+        A: Borrow<K>,
+        B: Borrow<V>,
+        I: ExactSizeIterator<Item = (A, B)>,
```

Comment:
You shouldn't have to require an `ExactSizeIterator`. You can use `.size_hint()` to preallocate the containers, if the size hint is exact that's nice, if it's not then it's not the end of the world either. You can then make this implement `FromIterator`.

Reply:
No this isn't for preallocating the container. We need the exact size for hash computation `hasher.finish() as usize % m`.

Reply:
But you have that after collecting into `K::Container` even if the iterator wasn't exact.

Reply:
Changed to `Iterator`.